src/main/java/jcifs/smb/SmbSessionImpl.java

```diff
@@ -0,0 +1,1297 @@
+                }
+                catch ( Exception se ) {
+                    log.debug("Session setup failed", se);
+                    if ( this.connectionState.compareAndSet(1, 0) ) {
+                        // only try to logoff if we have not completed the session setup, ignore errors from chained
+                        // responses
+                        logoff(true, true);
+                    }
+                    throw se;
+                }
+                finally {
+                    trans.notifyAll();
+                }
+            }
+        }
+    }
+
+
+    /**
+     * @param trans
+     * @param chain
+     * @param andxResponse
+     * @throws SmbException
+     */
+    @SuppressWarnings ( "unchecked" )
+    private <T extends CommonServerMessageBlock> T sessionSetupSMB2 ( SmbTransportImpl trans, final String tdomain,
+            ServerMessageBlock2Request<?> chain, T andxResponse ) throws CIFSException, GeneralSecurityException {
+        final Smb2NegotiateResponse negoResp = (Smb2NegotiateResponse) trans.getNegotiateResponse();
+        Smb2SessionSetupRequest request = null;
+        Smb2SessionSetupResponse response = null;
+        SmbException ex = null;
+        SSPContext ctx = null;
+        byte[] token = negoResp.getSecurityBlob();
+        final int securityMode = ( ( negoResp.getSecurityMode() & Smb2Constants.SMB2_NEGOTIATE_SIGNING_REQUIRED ) != 0 ) || trans.isSigningEnforced()
+                ? Smb2Constants.SMB2_NEGOTIATE_SIGNING_REQUIRED : Smb2Constants.SMB2_NEGOTIATE_SIGNING_ENABLED;
+        boolean anonymous = this.credentials.isAnonymous();
+        long sessId = 0;
+
+        boolean preauthIntegrity = negoResp.getSelectedDialect().atLeast(DialectVersion.SMB311);
+        this.preauthIntegrityHash = preauthIntegrity ? trans.getPreauthIntegrityHash() : null;
+
+        if ( this.preauthIntegrityHash != null && log.isDebugEnabled() ) {
+            log.debug("Initial session preauth hash " + Hexdump.toHexString(this.preauthIntegrityHash));
+        }
+
+        while ( true ) {
+            Subject s = this.credentials.getSubject();
+            if ( ctx == null ) {
+                ctx = createContext(trans, tdomain, negoResp, !anonymous, s);
+            }
+            token = createToken(ctx, token, s);
+
+            if ( token != null ) {
+                request = new Smb2SessionSetupRequest(this.getContext(), securityMode, negoResp.getCommonCapabilities(), 0, token);
+                // here, messages are rejected with NOT_SUPPORTED if we start signing as soon as we can, wait until
+                // session setup complete
+
+                request.setSessionId(sessId);
+                request.retainPayload();
+
+                try {
+                    response = trans.send(request, null, EnumSet.of(RequestParam.RETAIN_PAYLOAD));
+                    sessId = response.getSessionId();
+                }
+                catch ( SmbAuthException sae ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)